File: backend/app/llm/json_utils.py
from __future__ import annotations

import json
import re
from typing import TypeVar

from pydantic import BaseModel

T = TypeVar("T", bound=BaseModel)

_FENCE_START = re.compile(r"^```(?:json)?\s*", re.IGNORECASE)
_FENCE_END = re.compile(r"\s*```$")


def strip_json_fences(text: str) -> str:
    cleaned = text.strip()
    if cleaned.startswith("```"):
        cleaned = _FENCE_START.sub("", cleaned)
        cleaned = _FENCE_END.sub("", cleaned)
    return cleaned.strip()


def parse_model_from_text(schema: type[T], text: str) -> T:
    cleaned = strip_json_fences(text)
    try:
        return schema.model_validate_json(cleaned)
    except Exception:
        # Last resort: locate the outermost JSON object/array.
        start_obj = cleaned.find("{")
        start_arr = cleaned.find("[")
        starts = [index for index in (start_obj, start_arr) if index >= 0]
        if not starts:
            raise
        start = min(starts)
        end_obj = cleaned.rfind("}")
        end_arr = cleaned.rfind("]")
        end = max(end_obj, end_arr)
        if end <= start:
            raise
        return schema.model_validate(json.loads(cleaned[start : end + 1]))

File: backend/src/llm/json_utils.ts
import type { z } from "zod";

const FENCE_START = /^```(?:json)?\s*/i;
const FENCE_END = /\s*```$/;

export function stripJsonFences(text: string): string {
  let cleaned = text.trim();
  if (cleaned.startsWith("```")) {
    cleaned = cleaned.replace(FENCE_START, "");
    cleaned = cleaned.replace(FENCE_END, "");
  }
  return cleaned.trim();
}

export function parseModelFromText<T>(schema: z.ZodType<T>, text: string): T {
  const cleaned = stripJsonFences(text);
  try {
    return schema.parse(JSON.parse(cleaned));
  } catch (firstError) {
    const startObj = cleaned.indexOf("{");
    const startArr = cleaned.indexOf("[");
    const starts = [startObj, startArr].filter((index) => index >= 0);
    if (starts.length === 0) {
      throw firstError;
    }
    const start = Math.min(...starts);
    const endObj = cleaned.lastIndexOf("}");
    const endArr = cleaned.lastIndexOf("]");
    const end = Math.max(endObj, endArr);
    if (end <= start) {
      throw firstError;
    }
    return schema.parse(JSON.parse(cleaned.slice(start, end + 1)));
  }
}
